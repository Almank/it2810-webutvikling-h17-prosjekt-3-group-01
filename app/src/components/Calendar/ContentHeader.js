/**
 * Created by almank on 14.10.2017.
 */
import React from 'react';
import { View, Text, StyleSheet } from 'react-native';

export class ContentHeader extends React.Component {

    render(){
        return (
            <View style={styles.container}>
                <Text style={styles.item}>When</Text>
                <Text style={styles.item}>Title</Text>
                <Text style={styles.item}>About</Text>
            </View>
        )
    }
}

const styles = StyleSheet.create({
    container: {
        display: 'flex',
        marginTop: 20,
        flexDirection: 'row',
        justifyContent: 'center',
        flexWrap: 'wrap',
        borderBottomWidth:1.5,
        borderBottomColor: 'black',
    },
    item: {
        flexBasis:'33%',
        textAlign:'center',
        fontSize:18,
        marginBottom: 20,
    }
});
