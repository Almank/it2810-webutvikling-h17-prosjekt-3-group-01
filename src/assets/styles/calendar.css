.calendarContent {
    display: flex;
    flex-direction: column;
    margin: 0 10%;
}

.title {
    text-align: center;
    font-weight: 700;
    border-bottom:1px solid rgba(155,155,155,0.5);
    padding-bottom: 15px;
    margin-bottom: 5px;
}

.weekBox {
    display: flex;
    flex-flow: row wrap;
    justify-content: center;
    align-self: center;
}

.dayBox {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    flex-basis:100px;
    margin:15px;
    border: 1px solid rgba(155,155,155,0.5);
    background-color: #FFF;
    height: 100px;
}

.dayBox p {
    margin:0;
    font-size:11px;
}

.dayBox h1 {
    font-size:32px;
}

.dayBox:hover {
    border-color: #FF9505;
    cursor: pointer;
}

.dayDate {
    margin-top: 5px;
    margin-bottom: 0px;
}

.calendarBox {
    background-color: white;
    width: 100%;
    height:auto;
    margin:0px;
    padding: 20px;
}