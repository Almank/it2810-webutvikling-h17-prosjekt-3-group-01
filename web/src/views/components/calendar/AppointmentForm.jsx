/**
 * Created by almank on 14.10.2017.
 */
import React from 'react';

export class AppointmentForm extends React.Component {

    render(){
        return (
            <div className='formContainer' style={{
                //Need to style inline to remove having to double click button first time.
                display: 'none'
            }}>
                <button className='absolute-icon-top-right' onClick={this.props.closeForm}>
                    <span className='glyphicon glyphicon-remove'/>
                </button>
                <form className={'form'}>
                    <h3>Create new appointment</h3>
                    <label htmlFor="date">Date</label>
                    <input type='date' name='date' id='date' className="dateInput" required />
                    <label htmlFor="time">Time</label>
                    <input type='time' id='time' className="timeInput" required />
                    <label htmlFor="title">Title</label>
                    <input type="text" id='title' name="title" className="titleInput" maxLength='19' />
                    <label htmlFor="what">What</label>
                    <textarea type='text' id='what' className="textInput" maxLength='200'  />
                    <input type='submit' className='submitButton' onClick={this.props.submitForm}/>
                </form>
            </div>
        );
    }
}