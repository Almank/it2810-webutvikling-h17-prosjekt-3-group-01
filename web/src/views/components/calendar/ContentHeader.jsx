/**
 * Created by almank on 14.10.2017.
 */
import React from 'react';
import PropTypes from 'prop-types';

export class ContentHeader extends React.Component {

    render(){
        return (
            <div className='calendarBoxContent'>
                <div className='leftCalendarBoxContent'>
                    <h2>When</h2>
                </div>

                <div className='middleCalendarBoxContent'>
                    <h2>Title</h2>
                </div>
                <div className='rightCalendarBoxContent'>
                    <h2>About</h2>
                </div>
                <button className='icon-topright' onClick={this.props.closeForm}>
                    <span className='glyphicon glyphicon-plus'/>
                </button>
            </div>
        )
    }
}

ContentHeader.PropTypes = {
    closeForm: PropTypes.func,
};