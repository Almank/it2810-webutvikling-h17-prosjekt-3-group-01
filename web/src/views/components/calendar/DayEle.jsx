/**
 * Created by almank on 08.10.2017.
 */
import React from 'react';

export class DayEle extends React.Component {

    render(){
        return (
            <button key={this.props.dateFull} className={'dayBox'} onClick={this.props.change.bind(null, this)}>
                <p className={'dayName'}>{this.props.dateName}</p>
                <h1 className={'dayDate'}>{this.props.dayDate}</h1>
            </button>
        )
    }
}
