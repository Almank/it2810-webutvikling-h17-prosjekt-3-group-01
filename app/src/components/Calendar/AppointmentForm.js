/**
 * Created by almank on 14.10.2017.
 */
import React from 'react';
import {ContentHeader} from "./ContentHeader";
import { View, Text, TextInput, TouchableOpacity, Modal, StyleSheet } from 'react-native';

export class AppointmentForm extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            modalVisible: false,
        };
        this.submitInput = this.submitInput.bind(this);
    }

    //Send data to calendar on submit
    submitInput() {
        if (this.state.dateValue !== null && this.state.timeValue !== null && this.state.titleValue !== null){
            this.props.getValues([this.state.dateValue, this.state.timeValue, this.state.titleValue, this.state.textValue]);
        } else {
            this.setState({
                dateValue: new Date().toISOString().slice(0, 10),
                timeValue: null,
                titleValue: null,
                textValue: null,
            })
        }
    }

    //Hide/show form.
    setModalVisible(visible) {
        this.setState({modalVisible: visible});
    }

    render(){
        return (
            <View>
                <TouchableOpacity
                    onPress={() => {
                        this.setModalVisible(true)
                        }}
                        title='Press me'
                        style={[styles.addAppButton, styles.shadow]}>
                    <Text style={styles.textButton}>ADD APPOINTMENT</Text>
                </TouchableOpacity>
                <ContentHeader />
                <Modal
                    animationType="slide"
                    transparent={false}
                    visible={this.state.modalVisible}
                    onRequestClose={() => {alert("Form has been closed.")}}>
                    <View style={styles.formContainer}>
                        <View>
                            <Text style={styles.title}>Create new appointment</Text>
                        </View>
                        <Text htmlFor="date" style={styles.text}>Date - YYYY-MM-DD</Text>
                        <TextInput onChangeText={(dateValue) => this.setState({dateValue})}
                                   value={this.state.dateValue}
                                   placeholder={'YYYY-MM-DD'}
                                   style={[styles.inputField, styles.shadow]}/>
                        <Text htmlFor="time" style={styles.text}>Time - HH:MM</Text>
                        <TextInput onChangeText={(timeValue) => this.setState({timeValue})}
                                   value={this.state.timeValue}
                                   placeholder={'HH:MM'}
                                   style={[styles.inputField, styles.shadow]}/>
                        <Text htmlFor="title" style={styles.text}>Title</Text>
                        <TextInput onChangeText={(titleValue) => this.setState({titleValue})}
                                   value={this.state.titleValue}
                                   style={[styles.inputField, styles.shadow]}/>
                        <Text htmlFor="what" style={styles.text}>What</Text>
                        <TextInput onChangeText={(textValue) => this.setState({textValue})}
                                   value={this.state.textValue}
                                   style={[styles.inputField, styles.shadow]}/>
                        <TouchableOpacity onPress={() => {
                            this.setModalVisible(!this.state.modalVisible);
                            //To avoid getting stuck by having alert behind modal.
                            setTimeout(() => {
                                this.submitInput();
                            }, 100);

                            }}
                            title='Submit'
                            style={[styles.addAppButton, styles.shadow]}>
                            <Text style={styles.textButton}>Submit</Text>
                        </TouchableOpacity>
                        <TouchableOpacity onPress={() => {
                            this.setModalVisible(!this.state.modalVisible)
                        }}
                                          style={[styles.addAppButton, styles.shadow]}>
                            <Text style={styles.textButton}>Cancel</Text>
                        </TouchableOpacity>
                    </View>
                </Modal>
            </View>
        );
    }
}

const styles = StyleSheet.create({
    addAppButton: {
        marginTop: 20,
        marginLeft:'15%',
        marginRight:'15%',
        flexBasis: 40,
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center',
        backgroundColor: '#FF9505',
        borderWidth: 1,
        borderColor: 'rgba(155,155,155,0.5)',
    },
    textButton: {
        fontSize:16,
        color:'white',
    },
    formContainer: {
        display: 'flex',
        justifyContent: 'center',
        flexDirection: 'column',
        height:'100%',
    },
    inputField: {
        padding: 5,
        marginLeft: '15%',
        marginRight: '15%',
        flexBasis: 40,
        backgroundColor: 'white',
        borderWidth: 1,
        borderColor: 'rgba(155,155,155,0.5)',
    },
    shadow: {
        shadowColor: 'black',
        shadowOffset: {width:0, height: 2},
        shadowOpacity: 0.2,
        shadowRadius: 5,
    },
    text: {
        marginLeft: '15%',
        marginRight: '15%',
        marginBottom: 3,
        marginTop: 10,
    },
    title: {
        fontSize: 25,
        textAlign: 'center',
    }

});