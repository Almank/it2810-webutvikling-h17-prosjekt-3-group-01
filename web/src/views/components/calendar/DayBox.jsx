/**
 * Created by almank on 02.10.2017.
 */
import React from 'react';
import PropTypes from 'prop-types';
import {DayEle} from '../calendar/DayEle';

export class DayBox extends React.Component {
    constructor(props){
        super(props);
    }

    getDateFull(i){
        let temp = new Date();
        temp.setDate(temp.getDate() + i);
        return temp.toISOString().slice(0, 10);
    }

    getDayDate(i){
        let temp = new Date().getDate();
        return temp + i;
    }

    getDayName(i){
        const days = [ 'SUNDAY', 'MONDAY', 'TUESDAY', 'WEDNESDAY', 'THURSDAY', 'FRIDAY', 'SATURDAY'];
        let dayOfWeek = new Date().getDay();
        if (i + dayOfWeek >= 7)
            return days[i+dayOfWeek-7];
        else {
            return days[i+dayOfWeek];
        }
    }

    render(){
        let preset = ['TODAY', 'TOMORROW', 2, 3, 4, 5, 6];
        let weekdays = [];
        for(let i = 0; i < 7; i ++){
            let dayname = [];
            if(i < 2){
                dayname.push(preset[i]);
            } else {
                dayname.push(this.getDayName(preset[i]))
            }
            console.log(i);
            weekdays.push(<DayEle change={this.props.change}
                                  key={this.getDateFull(i)}
                                  dateFull={this.getDateFull(i)}
                                  dateName={ dayname }
                                  dayDate={this.getDayDate(i)} />)
        }

        return (
            <div className={'weekBox'}>
                { weekdays }
            </div>
        );
    }
}

DayBox.PropTypes = {
    change: PropTypes.string.isRequired,
};