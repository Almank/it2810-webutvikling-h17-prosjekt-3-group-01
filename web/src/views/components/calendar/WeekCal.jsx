/**
 * Created by almank on 02.10.2017.
 */
import React from 'react';
import PropTypes from 'prop-types';
import {DayEle} from '../calendar/DayEle';

export class WeekCal extends React.Component {
    getDateFull(i){
        let temp = new Date(Date.now() - new Date().getTimezoneOffset() * 60000);
        temp.setDate(temp.getDate() + i);
        return temp.toISOString().slice(0, 10);
    }

    getDayDate(i) {
        return new Date().getDate() + i;
    }

    getDayName(i) {
        const dayNames = ['MONDAY', 'TUESDAY', 'WEDNESDAY', 'THURSDAY', 'FRIDAY', 'SATURDAY', 'SUNDAY'];
        const firstTwoDays = ['TODAY', 'TOMORROW'];
        let weekDay = new Date().getDay();
        if (i < 2)
            return firstTwoDays[i];
        if (i + weekDay > 7)
            return dayNames[i + weekDay - 7 - 1]; //Minus 7 days, and minus 1 since it starts at 0.
        return dayNames[i + weekDay - 1];
    }

    render() {
        let weekdays = [];
        for (let i = 0; i < 7; i++) {
            weekdays.push(<DayEle change={this.props.change}
                                  key={this.getDateFull(i)}
                                  dateFull={this.getDateFull(i)}
                                  dateName={this.getDayName(i)}
                                  dayDate={this.getDayDate(i)}/>)
        }
        return (
            <div className={'weekBox'}>
                {weekdays}
            </div>
        );
    }
}

WeekCal.PropTypes = {
    change: PropTypes.string.isRequired,
};