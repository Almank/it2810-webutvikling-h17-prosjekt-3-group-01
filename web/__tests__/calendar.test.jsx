import React from "react";
import {WeekCal} from "../src/views/components/calendar/WeekCal";
import {DayEle} from "../src/views/components/calendar/DayEle";
import {ContentHeader} from "../src/views/components/calendar/ContentHeader";
import {AppointmentForm} from "../src/views/components/calendar/AppointmentForm";


// Variables used to test the source code
const WEEKDAYS = ['MONDAY', 'TUESDAY', 'WEDNESDAY', 'THURSDAY', 'FRIDAY', 'SATURDAY', 'SUNDAY'];


const SHALLOW_WEEK_CAL = shallow(<WeekCal/>);
const INST_WEEK_CAL = SHALLOW_WEEK_CAL.instance();

// Passing in a random function because we only care about it actually rendering, and not if the function is
// correct.
const RENDER_DAY_ELEM = render(<DayEle dateFull={"2017-01-28"} dateName={"Monday"} dateDay={"28"}
                                       change={() => console.log("Change")}/>);
const RENDER_CONTENT_HEADER = render(<ContentHeader closeForm={() => console.log("Close Form")}/>);

const RENDER_APPOINTMENT_FORM = render(<AppointmentForm closeForm={() => console.log("Close Form")}
                                                        openForm={() => console.log("Open Form")}/>);

const DATE = new Date();
let CURRENT_MONTH = DATE.getMonth() + 1;
const DAYS_IN_MONTH = new Date(DATE.getFullYear(), CURRENT_MONTH, 0).getDate();
const DAYS_LEFT_IN_MONTH = DAYS_IN_MONTH - DATE.getDate();


describe("CALENDAR:", () => {
    describe('WeekCal', () => {
        it("Should render(shallow) correctly", () => {
            expect(SHALLOW_WEEK_CAL).toMatchSnapshot();
        });

        it('Should render 7 <DayEle> components', () => {
            expect(SHALLOW_WEEK_CAL.find(DayEle).length).toBe(7);
        });

        describe('WeekCal.getDayName()', () => {
            it('Should return "TODAY" given input=0 ', () => {
                expect(INST_WEEK_CAL.getDayName(0)).toBe('TODAY');
            });
            it('Should return "TOMORROW" given input=1', () => {
                expect(INST_WEEK_CAL.getDayName(1)).toBe("TOMORROW");
            });
            // The following test may be wrong, should come back to it later.
            // May be that the function does not handle looping of a week, which seems to be the case
            it("Should return correct weekday depending on input", () => {
                for (let i = 7; i < 14; i++) {
                    expect(INST_WEEK_CAL.getDayName(i)).toBe(WEEKDAYS[i + DATE.getDay() - 7 - 1]);
                }
            });
        });
        describe("WeekCal.getDayDate()", () => {
            it("Should return today's day DATE, given input= ", () => {
                expect(INST_WEEK_CAL.getDayDate(0)).toBe(DATE.getDate());
            });
            it("Should return correct day DATE, given input <= days left in month", () => {
                expect(INST_WEEK_CAL.getDayDate(DAYS_LEFT_IN_MONTH)).toBe(DAYS_IN_MONTH);
                expect(INST_WEEK_CAL.getDayDate(DAYS_LEFT_IN_MONTH - 1)).toBe(DAYS_IN_MONTH - 1);
            });
            it("Should return correct day DATE when input-offset exceeds days in the month", () => {
                expect(INST_WEEK_CAL.getDayDate(DAYS_LEFT_IN_MONTH + 1)).toBe(1);
            });
        });
        describe("WeekCal.getFullDate()", () => {
            it("Should return today's DATE on format [yyyy-mm-dd], given input=0 ", () => {
                expect(INST_WEEK_CAL.getDateFull(0)).toBe(DATE.getFullYear() + "-" + (DATE.getMonth() + 1) + "-" + DATE.getDate());
            });
            it("Should return correct DATE on format [yyy-mm-dd], given input <= days left in the month", () => {
                // Date().getMonth() returns a number between 0-11. Therefore we need to add 1 to the CURRENT_MONTH value
                // Also we need to handle wen the month is less than 10.
                let tempMonth = CURRENT_MONTH;
                if (tempMonth.length === 1) tempMonth = "0" + tempMonth;

                expect(INST_WEEK_CAL.getDateFull(DAYS_LEFT_IN_MONTH)).toBe(
                    DATE.getFullYear() + "-" + tempMonth + "-" + DAYS_IN_MONTH);
            });
            it("Should return correct DATE on format [yyyy-mm-dd] when input-offset exceeds days in the month", () => {
                // Added handling for months which will return a value less than 10, and the changing from December
                // to january
                let tempMonth = CURRENT_MONTH + 1;
                if (tempMonth.length === 1) tempMonth = "0" + tempMonth;
                if (tempMonth === 12) tempMonth = "01";
                expect(INST_WEEK_CAL.getDateFull(DAYS_LEFT_IN_MONTH + 1)).toBe(
                    DATE.getFullYear() + "-" + tempMonth + "-01");
            });
        });
    });
    describe("DayEle", () => {
        it("Should render correctly", () => {
            expect(RENDER_DAY_ELEM).toMatchSnapshot();
        });
        it("Should render 1 p-tag and 1 h1-tag", () => {
            expect(RENDER_DAY_ELEM.find("p").length).toBe(1);
            expect(RENDER_DAY_ELEM.find("h1").length).toBe(1);
        });
    });
    describe("ContentHeader", () => {
        it("Should render correctly", () => {
            expect(RENDER_CONTENT_HEADER).toMatchSnapshot();
        });
        it("Should render 3 h2s, 1 button and 1 span", () => {
            expect(RENDER_CONTENT_HEADER.find("h2").length).toBe(3);
            expect(RENDER_CONTENT_HEADER.find("button").length).toBe(1);
            expect(RENDER_CONTENT_HEADER.find("span").length).toBe(1);
        });
    });
    describe("AppointmentForm", () => {
        it("Should render correctly", () => {
            expect(RENDER_APPOINTMENT_FORM).toMatchSnapshot();
        });
    });

});