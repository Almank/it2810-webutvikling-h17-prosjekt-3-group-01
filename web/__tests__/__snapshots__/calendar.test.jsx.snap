// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CALENDAR: AppointmentForm Should render correctly 1`] = `
<div
  class="formContainer"
  style="display:none"
>
  <button
    class="absolute-icon-top-right"
  >
    <span
      class="glyphicon glyphicon-remove"
    />
  </button>
  <form
    class="form"
  >
    <h3>
      Create new appointment
    </h3>
    <label
      for="date"
    >
      Date
    </label>
    <input
      class="dateInput"
      id="date"
      name="date"
      required=""
      type="date"
    />
    <label
      for="time"
    >
      Time
    </label>
    <input
      class="timeInput"
      id="time"
      required=""
      type="time"
    />
    <label
      for="title"
    >
      Title
    </label>
    <input
      class="titleInput"
      id="title"
      maxlength="19"
      name="title"
      type="text"
    />
    <label
      for="what"
    >
      What
    </label>
    <textarea
      class="textInput"
      id="what"
      maxlength="200"
      type="text"
    />
    <input
      class="submitButton"
      type="submit"
    />
  </form>
</div>
`;

exports[`CALENDAR: ContentHeader Should render correctly 1`] = `
<div
  class="calendarBoxContent"
>
  <div
    class="leftCalendarBoxContent"
  >
    <h2>
      When
    </h2>
  </div>
  <div
    class="middleCalendarBoxContent"
  >
    <h2>
      Title
    </h2>
  </div>
  <div
    class="rightCalendarBoxContent"
  >
    <h2>
      About
    </h2>
  </div>
  <button
    class="icon-topright"
  >
    <span
      class="glyphicon glyphicon-plus"
    />
  </button>
</div>
`;

exports[`CALENDAR: DayEle Should render correctly 1`] = `
<button
  class="dayBox"
>
  <p
    class="dayName"
  >
    Monday
  </p>
  <h1
    class="dayDate"
  />
</button>
`;

exports[`CALENDAR: WeekCal Should render(shallow) correctly 1`] = `
<div
  className="weekBox"
>
  <DayEle
    dateFull="2017-10-17"
    dateName="TODAY"
    dayDate={17}
    key="2017-10-17"
  />
  <DayEle
    dateFull="2017-10-18"
    dateName="TOMORROW"
    dayDate={18}
    key="2017-10-18"
  />
  <DayEle
    dateFull="2017-10-19"
    dateName="THURSDAY"
    dayDate={19}
    key="2017-10-19"
  />
  <DayEle
    dateFull="2017-10-20"
    dateName="FRIDAY"
    dayDate={20}
    key="2017-10-20"
  />
  <DayEle
    dateFull="2017-10-21"
    dateName="SATURDAY"
    dayDate={21}
    key="2017-10-21"
  />
  <DayEle
    dateFull="2017-10-22"
    dateName="SUNDAY"
    dayDate={22}
    key="2017-10-22"
  />
  <DayEle
    dateFull="2017-10-23"
    dateName="MONDAY"
    dayDate={23}
    key="2017-10-23"
  />
</div>
`;
